Flaky testParallelStressMultipleThreadsMultipleWorkersPositiveLimitJitDisabled

### Describe the bug

Possible bug in async offload:

```
2022-09-09T13:13:31.9157982Z 2022-09-09T13:13:31.882773Z I i.q.t.t.TestUtils cleaned [worker=1]
2022-09-09T13:13:31.9829710Z java.lang.AssertionError: expected:<v
2022-09-09T13:13:31.9830600Z 3394168647660478011
2022-09-09T13:13:31.9831163Z 4086802474270249591
2022-09-09T13:13:31.9831645Z 3958193676455060057
2022-09-09T13:13:31.9832130Z 3619114107112892010
2022-09-09T13:13:31.9832691Z 3705833798044144433
2022-09-09T13:13:31.9833167Z 4238042693748641409
2022-09-09T13:13:31.9833705Z 3518554007419864093
2022-09-09T13:13:31.9834721Z 4014104627539596639
2022-09-09T13:13:31.9835293Z 3393210801760647293
2022-09-09T13:13:31.9835869Z 4099611147050818391
2022-09-09T13:13:31.9836421Z > but was:<v
2022-09-09T13:13:31.9836991Z 3394168647660478011
2022-09-09T13:13:31.9837565Z 4086802474270249591
2022-09-09T13:13:31.9838133Z 3958193676455060057
2022-09-09T13:13:31.9838715Z 3619114107112892010
2022-09-09T13:13:31.9839281Z 3705833798044144433
2022-09-09T13:13:31.9839854Z 4238042693748641409
2022-09-09T13:13:31.9840421Z 3518554007419864093
2022-09-09T13:13:31.9841433Z 4014104627539596639
2022-09-09T13:13:31.9842011Z 3393210801760647293
2022-09-09T13:13:31.9842577Z 3944678179613436885
2022-09-09T13:13:31.9843094Z >
2022-09-09T13:13:31.9843707Z 	at org.junit.Assert.fail(Assert.java:89)
2022-09-09T13:13:31.9844454Z 	at org.junit.Assert.failNotEquals(Assert.java:835)
2022-09-09T13:13:31.9845213Z 	at org.junit.Assert.assertEquals(Assert.java:120)
2022-09-09T13:13:31.9846023Z 	at io.questdb.test.tools.TestUtils.assertEquals(TestUtils.java:331)
2022-09-09T13:13:31.9846875Z 	at io.questdb.test.tools.TestUtils.assertEquals(TestUtils.java:309)
2022-09-09T13:13:31.9847764Z 	at io.questdb.griffin.AbstractGriffinTest.assertCursor(AbstractGriffinTest.java:145)
2022-09-09T13:13:31.9848667Z 	at io.questdb.griffin.AsyncOffloadTest.assertCursor(AsyncOffloadTest.java:266)
2022-09-09T13:13:31.9850071Z 	at io.questdb.griffin.AsyncOffloadTest.assertQuery(AsyncOffloadTest.java:233)
2022-09-09T13:13:31.9851035Z 	at io.questdb.griffin.AsyncOffloadTest.lambda$testParallelStress$1(AsyncOffloadTest.java:379)
```


```
2022-09-09T13:13:32.1135092Z 2022-09-09T13:13:32.086369Z E i.q.TestListener ***** Test Failed *****io.questdb.griffin.AsyncOffloadTest.testParallelStressMultipleThreadsMultipleWorkersPositiveLimitJitDisabled : 
2022-09-09T13:13:32.1136532Z java.lang.AssertionError: expected:<0> but was:<1>
2022-09-09T13:13:32.1137280Z 	at org.junit.Assert.fail(Assert.java:89)
2022-09-09T13:13:32.1138034Z 	at org.junit.Assert.failNotEquals(Assert.java:835)
2022-09-09T13:13:32.1138806Z 	at org.junit.Assert.assertEquals(Assert.java:647)
2022-09-09T13:13:32.1139590Z 	at org.junit.Assert.assertEquals(Assert.java:633)
2022-09-09T13:13:32.1140457Z 	at io.questdb.griffin.AsyncOffloadTest.lambda$testParallelStress$2(AsyncOffloadTest.java:394)
2022-09-09T13:13:32.1141312Z 	at io.questdb.test.tools.TestUtils.execute(TestUtils.java:851)
2022-09-09T13:13:32.1142120Z 	at io.questdb.test.tools.TestUtils.execute(TestUtils.java:867)
2022-09-09T13:13:32.1143022Z 	at io.questdb.griffin.AsyncOffloadTest.testParallelStress(AsyncOffloadTest.java:347)
2022-09-09T13:13:32.1144124Z 	at io.questdb.griffin.AsyncOffloadTest.testParallelStressMultipleThreadsMultipleWorkersPositiveLimitJitDisabled(AsyncOffloadTest.java:168)
2022-09-09T13:13:32.1145281Z 	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2022-09-09T13:13:32.1146168Z 	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2022-09-09T13:13:32.1147128Z 	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2022-09-09T13:13:32.1147949Z 	at java.lang.reflect.Method.invoke(Method.java:566)
2022-09-09T13:13:32.1148803Z 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
2022-09-09T13:13:32.1149745Z 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2022-09-09T13:13:32.1150674Z 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
2022-09-09T13:13:32.1151679Z 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2022-09-09T13:13:32.1152709Z 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
2022-09-09T13:13:32.1153652Z 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
2022-09-09T13:13:32.1154464Z 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
2022-09-09T13:13:32.1155322Z 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
2022-09-09T13:13:32.1156323Z 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
2022-09-09T13:13:32.1157246Z 	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
2022-09-09T13:13:32.1158005Z 	at java.lang.Thread.run(Thread.java:829)
```

[86.txt.zip](https://github.com/questdb/questdb/files/9535895/86.txt.zip)


### To reproduce

_No response_

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**: 6.5.3-SNAPSHOT
- **OS**: ARM
- **Browser**:
```


### Additional context

_No response_

--- a/core/src/main/java/io/questdb/cairo/sql/RecordCursorFactory.java
+++ b/core/src/main/java/io/questdb/cairo/sql/RecordCursorFactory.java
@@ -29,7 +29,7 @@ import io.questdb.griffin.PlanSink;
 import io.questdb.griffin.Plannable;
 import io.questdb.griffin.SqlException;
 import io.questdb.griffin.SqlExecutionContext;
-import io.questdb.mp.Sequence;
+import io.questdb.mp.SCSequence;
 import io.questdb.std.Sinkable;
 import io.questdb.std.str.CharSink;
 
@@ -65,7 +65,7 @@ public interface RecordCursorFactory extends Closeable, Sinkable, Plannable {
         return null;
     }
 
-    default PageFrameSequence<?> execute(SqlExecutionContext executionContext, Sequence collectSubSeq, int order) throws SqlException {
+    default PageFrameSequence<?> execute(SqlExecutionContext executionContext, SCSequence collectSubSeq, int order) throws SqlException {
         return null;
     }
 

--- a/core/src/main/java/io/questdb/cairo/sql/async/PageFrameReduceJob.java
+++ b/core/src/main/java/io/questdb/cairo/sql/async/PageFrameReduceJob.java
@@ -219,7 +219,7 @@ public class PageFrameReduceJob implements Job, Closeable {
         if (!circuitBreaker.checkIfTripped(frameSequence.getStartTime(), frameSequence.getCircuitBreakerFd())) {
             record.of(frameSequence.getSymbolTableSource(), frameSequence.getPageAddressCache());
             record.setFrameIndex(task.getFrameIndex());
-            assert frameSequence.doneLatch.getCount() == 0;
+            assert !frameSequence.done;
             frameSequence.getReducer().reduce(workerId, record, task, circuitBreaker, stealingFrameSequence);
         } else {
             frameSequence.cancel();

--- a/core/src/main/java/io/questdb/cairo/sql/async/PageFrameReduceTask.java
+++ b/core/src/main/java/io/questdb/cairo/sql/async/PageFrameReduceTask.java
@@ -41,6 +41,7 @@ public class PageFrameReduceTask implements Closeable {
     private final DirectLongList rows;
     private int frameIndex = Integer.MAX_VALUE;
     private PageFrameSequence<?> frameSequence;
+    private long frameSequenceId;
 
     public PageFrameReduceTask(CairoConfiguration configuration) {
         this.rows = new DirectLongList(configuration.getPageFrameReduceRowIdListCapacity(), MemoryTag.NATIVE_OFFLOAD);
@@ -75,6 +76,10 @@ public class PageFrameReduceTask implements Closeable {
         return (PageFrameSequence<T>) frameSequence;
     }
 
+    public long getFrameSequenceId() {
+        return frameSequenceId;
+    }
+
     public PageAddressCache getPageAddressCache() {
         return frameSequence.getPageAddressCache();
     }
@@ -85,6 +90,7 @@ public class PageFrameReduceTask implements Closeable {
 
     public void of(PageFrameSequence<?> frameSequence, int frameIndex) {
         this.frameSequence = frameSequence;
+        this.frameSequenceId = frameSequence.getId();
         this.frameIndex = frameIndex;
         rows.clear();
     }

--- a/core/src/main/java/io/questdb/cairo/sql/async/PageFrameSequence.java
+++ b/core/src/main/java/io/questdb/cairo/sql/async/PageFrameSequence.java
@@ -100,8 +100,17 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
 
         final MCSequence pageFrameReduceSubSeq = messageBus.getPageFrameReduceSubSeq(shard);
         while (!done) {
+            if (dispatchStartFrameIndex == collectedFrameIndex + 1) {
+                // We know that all frames were collected. We're almost done.
+                if (!done) {
+                    // Looks like not all the frames were dispatched, so no one reached the very last frame and
+                    // reset the sequence via calling PageFrameReduceTask#collected(). Let's do it ourselves.
+                    reset();
+                }
+                break;
+            }
+
             // We were asked to steal work from the reduce queue and beyond, as much as we can.
-            final boolean allFramesReduced = (reduceCounter.get() == dispatchStartFrameIndex);
             boolean nothingProcessed = true;
             try {
                 nothingProcessed = PageFrameReduceJob.consumeQueue(reduceQueue, pageFrameReduceSubSeq, record, circuitBreaker, this);
@@ -118,17 +127,11 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
                     // Discard collected items.
                     final PageFrameReduceTask task = reduceQueue.get(cursor);
                     if (task.getFrameSequence() == this) {
+                        assert id == task.getFrameSequenceId() : "ids mismatch: " + id + ", " + task.getFrameSequenceId();
+                        collectedFrameIndex = task.getFrameIndex();
                         task.collected(true);
                     }
                     collectSubSeq.done(cursor);
-                } else if (cursor == -1 && allFramesReduced) {
-                    // The collect queue is empty while we know that all frames were reduced. We're almost done.
-                    if (!done) {
-                        // Looks like not all the frames were dispatched, so no one reached the very last frame and
-                        // reset the sequence via calling PageFrameReduceTask#collected(). Let's do it ourselves.
-                        reset();
-                    }
-                    break;
                 } else {
                     Os.pause();
                 }
@@ -176,6 +179,11 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
 
     public void collect(long cursor, boolean forceCollect) {
         assert cursor > -1;
+        LOG.debug()
+                .$("collecting [shard=").$(shard)
+                .$(", id=").$(id)
+                .$(", cursor=").$(cursor)
+                .I$();
         if (cursor == LOCAL_TASK_CURSOR) {
             collectedFrameIndex = localTask.getFrameIndex();
             localTask.collected();
@@ -256,8 +264,8 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
      * instead, should use getTask and collect methods. <code>Long.MAX_VALUE</code> is the
      * reserved cursor value for the local reduce task case.
      *
-     * @return the next cursor value or one of -1 and -2 values if the cursor failed and the
-     * caller should retry
+     * @return the next cursor value or -1 value if the cursor failed and the caller
+     * should retry
      */
     public long next() {
         assert collectedFrameIndex < frameCount - 1;
@@ -267,6 +275,7 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
                 PageFrameReduceTask task = reduceQueue.get(cursor);
                 PageFrameSequence<?> thatFrameSequence = task.getFrameSequence();
                 if (thatFrameSequence == this) {
+                    assert id == task.getFrameSequenceId() : "ids mismatch: " + id + ", " + task.getFrameSequenceId();
                     return cursor;
                 } else {
                     // Not our task, nothing to collect. Go for another spin.
@@ -368,7 +377,7 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
     }
 
     /**
-     * This method is re enterable. It has to be in case queue capacity is smaller than number of frames to
+     * This method is re-enterable. It has to be in case queue capacity is smaller than number of frames to
      * be dispatched. When it is the case, frame count published so far is stored in the `frameSequence`.
      * This method has no responsibility to deal with "collect" stage hence it deals with everything to
      * unblock the collect stage.

--- a/core/src/main/java/io/questdb/cairo/sql/async/PageFrameSequence.java
+++ b/core/src/main/java/io/questdb/cairo/sql/async/PageFrameSequence.java
@@ -179,11 +179,6 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
 
     public void collect(long cursor, boolean forceCollect) {
         assert cursor > -1;
-        LOG.debug()
-                .$("collecting [shard=").$(shard)
-                .$(", id=").$(id)
-                .$(", cursor=").$(cursor)
-                .I$();
         if (cursor == LOCAL_TASK_CURSOR) {
             collectedFrameIndex = localTask.getFrameIndex();
             localTask.collected();

--- a/core/src/main/java/io/questdb/cairo/sql/async/PageFrameSequence.java
+++ b/core/src/main/java/io/questdb/cairo/sql/async/PageFrameSequence.java
@@ -31,7 +31,10 @@ import io.questdb.griffin.SqlException;
 import io.questdb.griffin.SqlExecutionContext;
 import io.questdb.log.Log;
 import io.questdb.log.LogFactory;
-import io.questdb.mp.*;
+import io.questdb.mp.MCSequence;
+import io.questdb.mp.MPSequence;
+import io.questdb.mp.RingQueue;
+import io.questdb.mp.SCSequence;
 import io.questdb.std.*;
 import io.questdb.std.datetime.millitime.MillisecondClock;
 
@@ -45,7 +48,6 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
     private static final AtomicLong ID_SEQ = new AtomicLong();
     private static final long LOCAL_TASK_CURSOR = Long.MAX_VALUE;
     private static final Log LOG = LogFactory.getLog(PageFrameSequence.class);
-    public final SOUnboundedCountDownLatch doneLatch = new SOUnboundedCountDownLatch();
     private final MillisecondClock clock;
     private final LongList frameRowCounts = new LongList();
     private final WeakClosableObjectPool<PageFrameReduceTask> localTaskPool;
@@ -54,10 +56,11 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
     private final AtomicInteger reduceCounter = new AtomicInteger(0);
     private final PageFrameReducer reducer;
     private final AtomicBoolean valid = new AtomicBoolean(true);
+    public volatile boolean done;
     private T atom;
     private SqlExecutionCircuitBreaker circuitBreaker;
     private long circuitBreakerFd;
-    private Sequence collectSubSeq;
+    private SCSequence collectSubSeq;
     private int collectedFrameIndex = -1;
     private int dispatchStartFrameIndex;
     private int frameCount;
@@ -96,9 +99,9 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
                 .I$();
 
         final MCSequence pageFrameReduceSubSeq = messageBus.getPageFrameReduceSubSeq(shard);
-        while (doneLatch.getCount() == 0) {
+        while (!done) {
             // We were asked to steal work from the reduce queue and beyond, as much as we can.
-            final boolean allFramesReduced = reduceCounter.get() == dispatchStartFrameIndex;
+            final boolean allFramesReduced = (reduceCounter.get() == dispatchStartFrameIndex);
             boolean nothingProcessed = true;
             try {
                 nothingProcessed = PageFrameReduceJob.consumeQueue(reduceQueue, pageFrameReduceSubSeq, record, circuitBreaker, this);
@@ -112,15 +115,15 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
             if (nothingProcessed) {
                 long cursor = collectSubSeq.next();
                 if (cursor > -1) {
-                    // Discard collect items.
-                    final PageFrameReduceTask tsk = reduceQueue.get(cursor);
-                    if (tsk.getFrameSequence() == this) {
-                        tsk.collected(true);
+                    // Discard collected items.
+                    final PageFrameReduceTask task = reduceQueue.get(cursor);
+                    if (task.getFrameSequence() == this) {
+                        task.collected(true);
                     }
                     collectSubSeq.done(cursor);
                 } else if (cursor == -1 && allFramesReduced) {
                     // The collect queue is empty while we know that all frames were reduced. We're almost done.
-                    if (doneLatch.getCount() == 0) {
+                    if (!done) {
                         // Looks like not all the frames were dispatched, so no one reached the very last frame and
                         // reset the sequence via calling PageFrameReduceTask#collected(). Let's do it ourselves.
                         reset();
@@ -291,11 +294,10 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
     public PageFrameSequence<T> of(
             RecordCursorFactory base,
             SqlExecutionContext executionContext,
-            Sequence collectSubSeq,
+            SCSequence collectSubSeq,
             T atom,
             int order
     ) throws SqlException {
-
         this.sqlExecutionContext = executionContext;
         this.startTime = clock.getTicks();
         this.circuitBreakerFd = executionContext.getCircuitBreaker().getFd();
@@ -337,8 +339,8 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
     public void reset() {
         // prepare to resend the same sequence as it might be required by toTop()
         frameRowCounts.clear();
-        assert doneLatch.getCount() == 0;
-        doneLatch.countDown();
+        assert !done;
+        done = true;
     }
 
     /**
@@ -347,15 +349,17 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
      */
     public void toTop() {
         if (frameCount > 0) {
+            long newId = ID_SEQ.incrementAndGet();
             LOG.debug().$("toTop [shard=").$(shard)
                     .$(", id=").$(id)
+                    .$(", newId=").$(newId)
                     .I$();
 
             await();
 
-            // done latch is reset by method call above
-            doneLatch.reset();
-            id = ID_SEQ.incrementAndGet();
+            // done is reset by method call above
+            done = false;
+            id = newId;
             dispatchStartFrameIndex = 0;
             collectedFrameIndex = -1;
             reduceCounter.set(0);
@@ -458,10 +462,10 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
             int frameCount,
             SymbolTableSource symbolTableSource,
             T atom,
-            Sequence collectSubSeq
+            SCSequence collectSubSeq
     ) {
         this.id = ID_SEQ.incrementAndGet();
-        this.doneLatch.reset();
+        this.done = false;
         this.valid.set(true);
         this.reduceCounter.set(0);
         this.shard = rnd.nextInt(messageBus.getPageFrameReduceShardCount());
@@ -510,6 +514,13 @@ public class PageFrameSequence<T extends StatefulAtom> implements Closeable {
         localTask.of(this, dispatchStartFrameIndex++);
 
         try {
+            LOG.debug()
+                    .$("reducing locally [shard=").$(shard)
+                    .$(", id=").$(id)
+                    .$(", frameIndex=").$(localTask.getFrameIndex())
+                    .$(", frameCount=").$(frameCount)
+                    .$(", active=").$(isActive())
+                    .I$();
             if (isActive()) {
                 PageFrameReduceJob.reduce(record, circuitBreaker, localTask, this, this);
             }

--- a/core/src/main/java/io/questdb/griffin/engine/table/AsyncFilteredRecordCursor.java
+++ b/core/src/main/java/io/questdb/griffin/engine/table/AsyncFilteredRecordCursor.java
@@ -51,12 +51,12 @@ class AsyncFilteredRecordCursor implements RecordCursor {
     private long frameRowIndex;
     private PageFrameSequence<?> frameSequence;
     private boolean isOpen;
-    // the OG rows remaining, used to reset the counter when re-running cursor from top();
+    // The OG rows remaining, used to reset the counter when re-running cursor from top().
     private long ogRowsRemaining;
     private PageAddressCacheRecord recordB;
     private DirectLongList rows;
     // Artificial limit on remaining rows to be returned from this cursor.
-    // It is typically copied from LIMIT clause on SQL statement
+    // It is typically copied from LIMIT clause on SQL statement.
     private long rowsRemaining;
 
     public AsyncFilteredRecordCursor(Function filter, boolean hasDescendingOrder) {
@@ -72,6 +72,7 @@ class AsyncFilteredRecordCursor implements RecordCursor {
                     .$("closing [shard=").$(frameSequence.getShard())
                     .$(", frameIndex=").$(frameIndex)
                     .$(", frameCount=").$(frameLimit)
+                    .$(", frameId=").$(frameSequence.getId())
                     .$(", cursor=").$(cursor)
                     .I$();
 
@@ -112,24 +113,29 @@ class AsyncFilteredRecordCursor implements RecordCursor {
 
     @Override
     public boolean hasNext() {
-        // check for the first hasNext call
+        // Check for the first hasNext call.
         if (frameIndex == -1 && frameLimit > -1) {
             fetchNextFrame();
         }
 
-        // we have rows in the current frame we still need to dispatch
+        // Check for already reached row limit.
+        if (rowsRemaining < 0) {
+            return false;
+        }
+
+        // We have rows in the current frame we still need to dispatch
         if (frameRowIndex < frameRowCount) {
             record.setRowIndex(rows.get(rowIndex()));
             frameRowIndex++;
             return checkLimit();
         }
 
-        // Release previous queue item.
+        // Release the previous queue item.
         // There is no identity check here because this check
-        // had been done when 'cursor' was assigned
+        // had been done when 'cursor' was assigned.
         collectCursor(false);
 
-        // do we have more frames?
+        // Do we have more frames?
         if (frameIndex < frameLimit) {
             fetchNextFrame();
             if (frameRowCount > 0 && frameRowIndex < frameRowCount) {
@@ -163,10 +169,11 @@ class AsyncFilteredRecordCursor implements RecordCursor {
 
     @Override
     public void toTop() {
-        // check if we at the top already and there is nothing to do
+        // Check if we at the top already and there is nothing to do.
         if (frameIndex == 0 && frameRowIndex == 0) {
             return;
         }
+        collectCursor(false);
         filter.toTop();
         frameSequence.toTop();
         rowsRemaining = ogRowsRemaining;
@@ -204,6 +211,7 @@ class AsyncFilteredRecordCursor implements RecordCursor {
                             .$("collected [shard=").$(frameSequence.getShard())
                             .$(", frameIndex=").$(task.getFrameIndex())
                             .$(", frameCount=").$(frameSequence.getFrameCount())
+                            .$(", frameId=").$(frameSequence.getId())
                             .$(", active=").$(frameSequence.isActive())
                             .$(", cursor=").$(cursor)
                             .I$();
@@ -216,7 +224,8 @@ class AsyncFilteredRecordCursor implements RecordCursor {
                         record.setFrameIndex(task.getFrameIndex());
                         break;
                     } else {
-                        this.frameRowCount = 0; // force reset frame size if frameSequence was canceled or failed
+                        // Force reset frame size if frameSequence was canceled or failed.
+                        this.frameRowCount = 0;
                         collectCursor(false);
                     }
                 } else {

--- a/core/src/main/java/io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java
+++ b/core/src/main/java/io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java
@@ -34,7 +34,6 @@ import io.questdb.cairo.sql.async.PageFrameSequence;
 import io.questdb.griffin.SqlException;
 import io.questdb.griffin.SqlExecutionContext;
 import io.questdb.mp.SCSequence;
-import io.questdb.mp.Sequence;
 import io.questdb.std.*;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -88,7 +87,7 @@ public class AsyncFilteredRecordCursorFactory extends AbstractRecordCursorFactor
     }
 
     @Override
-    public PageFrameSequence<AsyncFilterAtom> execute(SqlExecutionContext executionContext, Sequence collectSubSeq, int order) throws SqlException {
+    public PageFrameSequence<AsyncFilterAtom> execute(SqlExecutionContext executionContext, SCSequence collectSubSeq, int order) throws SqlException {
         return frameSequence.of(base, executionContext, collectSubSeq, filterAtom, order);
     }
 

--- a/core/src/main/java/io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java
+++ b/core/src/main/java/io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java
@@ -39,7 +39,6 @@ import io.questdb.griffin.SqlExecutionContext;
 import io.questdb.griffin.engine.functions.bind.CompiledFilterSymbolBindVariable;
 import io.questdb.jit.CompiledFilter;
 import io.questdb.mp.SCSequence;
-import io.questdb.mp.Sequence;
 import io.questdb.std.*;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -106,7 +105,7 @@ public class AsyncJitFilteredRecordCursorFactory extends AbstractRecordCursorFac
     }
 
     @Override
-    public PageFrameSequence<AsyncJitFilterAtom> execute(SqlExecutionContext executionContext, Sequence collectSubSeq, int order) throws SqlException {
+    public PageFrameSequence<AsyncJitFilterAtom> execute(SqlExecutionContext executionContext, SCSequence collectSubSeq, int order) throws SqlException {
         return frameSequence.of(base, executionContext, collectSubSeq, filterAtom, order);
     }
 

--- a/core/src/main/java/io/questdb/log/LogFactory.java
+++ b/core/src/main/java/io/questdb/log/LogFactory.java
@@ -50,7 +50,7 @@ public class LogFactory implements Closeable {
     public static final String DEFAULT_CONFIG_NAME = "log.conf";
     // placeholder that can be used in log.conf to point to $root/log/ dir
     public static final String LOG_DIR_VAR = "${log.dir}";
-    // name of default logging configuration file (in jar and in $root/conf/ dir )
+    // name of default logging configuration file (in jar and in $root/conf/ dir)
     private static final String DEFAULT_CONFIG = "/io/questdb/site/conf/" + DEFAULT_CONFIG_NAME;
     private static final int DEFAULT_LOG_LEVEL = LogLevel.INFO | LogLevel.ERROR | LogLevel.CRITICAL | LogLevel.ADVISORY;
     private static final int DEFAULT_MSG_SIZE = 4 * 1024;

--- a/core/src/main/java/io/questdb/mp/AbstractSSequence.java
+++ b/core/src/main/java/io/questdb/mp/AbstractSSequence.java
@@ -38,8 +38,8 @@ abstract class AbstractSSequence extends AbstractSequence implements Sequence {
     @Override
     public void clear() {
         setBarrier(OpenBarrier.INSTANCE);
-        value = -1;
         cache = -1;
+        value = -1;
     }
 
     @Override

--- a/core/src/main/java/io/questdb/mp/AbstractSequence.java
+++ b/core/src/main/java/io/questdb/mp/AbstractSequence.java
@@ -52,6 +52,7 @@ class LhsPadding {
     protected long p1, p2, p3, p4, p5, p6, p7;
 }
 
+@SuppressWarnings("unused")
 class RhsPadding extends Value {
     protected long p9, p10, p11, p12, p13, p14;
 

--- a/core/src/main/java/io/questdb/mp/FanOut.java
+++ b/core/src/main/java/io/questdb/mp/FanOut.java
@@ -51,7 +51,7 @@ public class FanOut implements Barrier {
 
     public FanOut and(Barrier barrier) {
         Holder _new;
-        boolean barrierNotSetUp = true;
+        Barrier root = null;
 
         final long current = this.barrier != null ? this.barrier.current() : -1;
         Unsafe.getUnsafe().loadFence();
@@ -64,14 +64,11 @@ public class FanOut implements Barrier {
                 return this;
             }
 
-            if (this.barrier != null) {
-                if (barrierNotSetUp) {
-                    Barrier b = barrier.root();
-                    b.setBarrier(this.barrier);
-                    b.setCurrent(current);
-                    barrierNotSetUp = false;
-                    Unsafe.getUnsafe().storeFence();
-                }
+            if (root == null && this.barrier != null) {
+                root = barrier.root();
+                root.setBarrier(this.barrier);
+                root.setCurrent(current);
+                Unsafe.getUnsafe().storeFence();
             }
             _new = new Holder();
             _new.barriers.addAll(h.barriers);
@@ -83,12 +80,15 @@ public class FanOut implements Barrier {
             }
             _new.setupWaitStrategy();
             if (Unsafe.getUnsafe().compareAndSwapObject(this, HOLDER, h, _new)) {
+                // catch up with others
+                if (root != null && this.barrier != null) {
+                    root.setCurrent(this.barrier.current());
+                }
                 break;
             }
         } while (true);
 
         Unsafe.getUnsafe().storeFence();
-
         return this;
     }
 
@@ -121,7 +121,6 @@ public class FanOut implements Barrier {
     }
 
     public void remove(Barrier barrier) {
-
         Unsafe.getUnsafe().storeFence();
 
         Holder _new;

--- a/core/src/main/java/io/questdb/mp/MCSequence.java
+++ b/core/src/main/java/io/questdb/mp/MCSequence.java
@@ -40,11 +40,6 @@ public class MCSequence extends AbstractMSequence {
         super(cycle, waitStrategy);
     }
 
-    public MCSequence(long value, int cycle) {
-        this(cycle);
-        this.value = value;
-    }
-
     public <T> void consumeAll(RingQueue<T> queue, QueueConsumer<T> consumer) {
         long cursor;
         do {

--- a/core/src/main/java/io/questdb/mp/SCSequence.java
+++ b/core/src/main/java/io/questdb/mp/SCSequence.java
@@ -88,9 +88,9 @@ public class SCSequence extends AbstractSSequence {
         return next0(next + 1);
     }
 
-    @Override
     // The method is final is because we call it from
     // the constructor.
+    @Override
     public final void setCurrent(long value) {
         this.value = value;
     }

--- a/core/src/test/java/io/questdb/mp/ConcurrentTest.java
+++ b/core/src/test/java/io/questdb/mp/ConcurrentTest.java
@@ -34,6 +34,7 @@ import java.util.Arrays;
 import java.util.concurrent.BrokenBarrierException;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
@@ -75,6 +76,135 @@ public class ConcurrentTest {
 
         latch.await();
         Assert.assertEquals(threads * iterations, doneCount.get());
+    }
+
+    @Test
+    public void testConcurrentFanOutAsyncOffloadPattern() {
+        // This test aims to reproduce async offload inter-thread communication pattern.
+
+        class Item {
+            int owner;
+            AtomicInteger reduceCounter;
+        }
+
+        final int capacity = 2;
+        final int collectorThreads = 4;
+        final int itemsToDispatch = 8;
+        final int iterations = 100;
+
+        final RingQueue<Item> queue = new RingQueue<Item>(() -> new Item(), capacity);
+        final MPSequence reducePubSeq = new MPSequence(capacity);
+        final MCSequence reduceSubSeq = new MCSequence(capacity);
+        final FanOut collectFanOut = new FanOut();
+
+        reducePubSeq.then(reduceSubSeq).then(collectFanOut).then(reducePubSeq);
+
+        final CyclicBarrier start = new CyclicBarrier(collectorThreads + 1);
+        final SOCountDownLatch latch = new SOCountDownLatch(collectorThreads + 1);
+        final AtomicInteger doneCollectors = new AtomicInteger();
+        final AtomicInteger anomalies = new AtomicInteger();
+
+        // Start reducer thread.
+        new Thread(() -> {
+            try {
+                start.await();
+
+                while (doneCollectors.get() != collectorThreads) {
+                    long cursor = reduceSubSeq.next();
+                    if (cursor > -1) {
+                        AtomicInteger reduceCounter = queue.get(cursor).reduceCounter;
+                        reduceSubSeq.done(cursor);
+                        reduceCounter.incrementAndGet();
+                    } else {
+                        Os.pause();
+                    }
+                }
+            } catch (InterruptedException | BrokenBarrierException e) {
+                e.printStackTrace();
+                anomalies.incrementAndGet();
+            } finally {
+                latch.countDown();
+            }
+        }).start();
+
+        // Start collector threads.
+        for (int i = 0; i < collectorThreads; i++) {
+            final int threadI = i;
+            new Thread(() -> {
+                final SCSequence collectSubSeq = new SCSequence();
+                final AtomicInteger reduceCounter = new AtomicInteger();
+                try {
+                    start.await();
+
+                    for (int j = 0; j < iterations; j++) {
+                        reduceCounter.set(0);
+                        collectFanOut.and(collectSubSeq);
+
+                        int collected = 0;
+
+                        // Dispatch all items.
+                        for (int k = 0; k < itemsToDispatch; k++) {
+                            while (true) {
+                                long dispatchCursor = reducePubSeq.next();
+                                if (dispatchCursor > -1) {
+                                    queue.get(dispatchCursor).owner = threadI;
+                                    queue.get(dispatchCursor).reduceCounter = reduceCounter;
+                                    reducePubSeq.done(dispatchCursor);
+                                    break;
+                                } else if (dispatchCursor == -1) {
+                                    // Collect as many items as we can.
+                                    long collectCursor;
+                                    while ((collectCursor = collectSubSeq.next()) > -1) {
+                                        if (queue.get(collectCursor).owner == threadI) {
+                                            queue.get(collectCursor).owner = -1;
+                                            queue.get(collectCursor).reduceCounter = null;
+                                            collected++;
+                                        }
+                                        collectSubSeq.done(collectCursor);
+                                    }
+                                    Os.pause();
+                                } else {
+                                    Os.pause();
+                                }
+                            }
+                        }
+
+                        while (reduceCounter.get() != itemsToDispatch) {
+                            Os.pause();
+                        }
+
+                        // Collect all remaining items.
+                        long collectCursor;
+                        while ((collectCursor = collectSubSeq.next()) > -1) {
+                            if (queue.get(collectCursor).owner == threadI) {
+                                queue.get(collectCursor).owner = -1;
+                                queue.get(collectCursor).reduceCounter = null;
+                                collected++;
+                            }
+                            collectSubSeq.done(collectCursor);
+                        }
+
+                        if (collected != itemsToDispatch) {
+                            anomalies.incrementAndGet();
+                        }
+
+                        collectFanOut.remove(collectSubSeq);
+                        collectSubSeq.clear();
+                    }
+                } catch (InterruptedException | BrokenBarrierException e) {
+                    e.printStackTrace();
+                    anomalies.incrementAndGet();
+                } finally {
+                    doneCollectors.incrementAndGet();
+                    latch.countDown();
+                }
+            }).start();
+        }
+
+        if (!latch.await(TimeUnit.SECONDS.toNanos(10))) {
+            Assert.fail("Wait limit exceeded");
+        }
+        Assert.assertEquals(0, anomalies.get());
     }
 
     @Test

--- a/core/src/test/java/io/questdb/mp/ConcurrentTest.java
+++ b/core/src/test/java/io/questdb/mp/ConcurrentTest.java
@@ -80,15 +80,15 @@ public class ConcurrentTest {
 
     @Test
     public void testConcurrentFanOutAsyncOffloadPattern() {
-        // This test aims to reproduce async offload inter-thread communication pattern.
+        // This test aims to reproduce a simplified variant of async offload inter-thread communication pattern.
 
         class Item {
             int owner;
-            AtomicInteger reduceCounter;
         }
 
         final int capacity = 2;
-        final int collectorThreads = 4;
+        final int collectorThreads = 2;
+        final int reducerThreads = 2;
         final int itemsToDispatch = 8;
         final int iterations = 100;
 
@@ -99,45 +99,44 @@ public class ConcurrentTest {
 
         reducePubSeq.then(reduceSubSeq).then(collectFanOut).then(reducePubSeq);
 
-        final CyclicBarrier start = new CyclicBarrier(collectorThreads + 1);
-        final SOCountDownLatch latch = new SOCountDownLatch(collectorThreads + 1);
+        final CyclicBarrier start = new CyclicBarrier(collectorThreads + reducerThreads);
+        final SOCountDownLatch latch = new SOCountDownLatch(collectorThreads + reducerThreads);
         final AtomicInteger doneCollectors = new AtomicInteger();
         final AtomicInteger anomalies = new AtomicInteger();
 
-        // Start reducer thread.
-        new Thread(() -> {
-            try {
-                start.await();
+        // Start reducer threads.
+        for (int i = 0; i < reducerThreads; i++) {
+            new Thread(() -> {
+                try {
+                    start.await();
 
-                while (doneCollectors.get() != collectorThreads) {
-                    long cursor = reduceSubSeq.next();
-                    if (cursor > -1) {
-                        AtomicInteger reduceCounter = queue.get(cursor).reduceCounter;
-                        reduceSubSeq.done(cursor);
-                        reduceCounter.incrementAndGet();
-                    } else {
-                        Os.pause();
+                    while (doneCollectors.get() != collectorThreads) {
+                        long cursor = reduceSubSeq.next();
+                        if (cursor > -1) {
+                            reduceSubSeq.done(cursor);
+                        } else {
+                            Os.pause();
+                        }
                     }
+                } catch (InterruptedException | BrokenBarrierException e) {
+                    e.printStackTrace();
+                    anomalies.incrementAndGet();
+                } finally {
+                    latch.countDown();
                 }
-            } catch (InterruptedException | BrokenBarrierException e) {
-                e.printStackTrace();
-                anomalies.incrementAndGet();
-            } finally {
-                latch.countDown();
-            }
-        }).start();
+            }).start();
+        }
 
         // Start collector threads.
         for (int i = 0; i < collectorThreads; i++) {
             final int threadI = i;
             new Thread(() -> {
                 final SCSequence collectSubSeq = new SCSequence();
-                final AtomicInteger reduceCounter = new AtomicInteger();
+
                 try {
                     start.await();
 
                     for (int j = 0; j < iterations; j++) {
-                        reduceCounter.set(0);
                         collectFanOut.and(collectSubSeq);
 
                         int collected = 0;
@@ -148,7 +147,6 @@ public class ConcurrentTest {
                                 long dispatchCursor = reducePubSeq.next();
                                 if (dispatchCursor > -1) {
                                     queue.get(dispatchCursor).owner = threadI;
-                                    queue.get(dispatchCursor).reduceCounter = reduceCounter;
                                     reducePubSeq.done(dispatchCursor);
                                     break;
                                 } else if (dispatchCursor == -1) {
@@ -157,7 +155,6 @@ public class ConcurrentTest {
                                     while ((collectCursor = collectSubSeq.next()) > -1) {
                                         if (queue.get(collectCursor).owner == threadI) {
                                             queue.get(collectCursor).owner = -1;
-                                            queue.get(collectCursor).reduceCounter = null;
                                             collected++;
                                         }
                                         collectSubSeq.done(collectCursor);
@@ -169,23 +166,18 @@ public class ConcurrentTest {
                             }
                         }
 
-                        while (reduceCounter.get() != itemsToDispatch) {
-                            Os.pause();
-                        }
-
-                        // Collect all remaining items.
-                        long collectCursor;
-                        while ((collectCursor = collectSubSeq.next()) > -1) {
-                            if (queue.get(collectCursor).owner == threadI) {
-                                queue.get(collectCursor).owner = -1;
-                                queue.get(collectCursor).reduceCounter = null;
-                                collected++;
+                        // Await for all remaining items.
+                        while (collected != itemsToDispatch) {
+                            long collectCursor = collectSubSeq.next();
+                            if (collectCursor > -1) {
+                                if (queue.get(collectCursor).owner == threadI) {
+                                    queue.get(collectCursor).owner = -1;
+                                    collected++;
+                                }
+                                collectSubSeq.done(collectCursor);
+                            } else {
+                                Os.pause();
                             }
-                            collectSubSeq.done(collectCursor);
-                        }
-
-                        if (collected != itemsToDispatch) {
-                            anomalies.incrementAndGet();
                         }
 
                         collectFanOut.remove(collectSubSeq);

--- a/core/src/test/java/io/questdb/mp/ConcurrentTest.java
+++ b/core/src/test/java/io/questdb/mp/ConcurrentTest.java
@@ -193,7 +193,7 @@ public class ConcurrentTest {
             }).start();
         }
 
-        if (!latch.await(TimeUnit.SECONDS.toNanos(10))) {
+        if (!latch.await(TimeUnit.SECONDS.toNanos(60))) {
             Assert.fail("Wait limit exceeded");
         }
         Assert.assertEquals(0, anomalies.get());

